Nobody is looking great today! Challenge crashes when no faces are found

Challenge crashes when no faces are found.

Reproducable by running challenge and blocking the kinect.

[INFO] [1515516760.313584]: 'You are all looking great today!            Keep looking in my camera!'
[INFO] [1515516761.748436]: getting one image...
[INFO] [1515516761.748780]: self._camera_last_image: None
[INFO] [1515516761.812574]: len(self._camera_last_image): 3507018
[INFO] [1515516761.812924]: got 3507018 bytes of image data
[INFO] [1515516762.923236]: found 0 face(s) in the image
[ERROR] [1515516762.923823]: InvalidUserCodeError: Could not execute state 'DETECT_CROWD' of type '<challenge_spr_states.detect.DetectCrowd object at 0x7f1ab2fa0610>': Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/smach/state_machine.py", line 247, in _update_once
    self._remappings[self._current_label]))
  File "/home/amigo/ros/kinetic/repos/https_/github.com/tue-robotics/tue_robocup.git/challenge_spr/src/challenge_spr_states/detect.py", line 30, in execute
    detections = self.recognize(tries)
  File "/home/amigo/ros/kinetic/repos/https_/github.com/tue-robotics/tue_robocup.git/challenge_spr/src/challenge_spr_states/detect.py", line 59, in recognize
    faces = img_cutout(best_image, img_recognitions_to_rois(best_detection))
  File "/home/amigo/ros/kinetic/system/src/robot_skills/src/robot_skills/util/image_operations.py", line 15, in img_recognitions_to_rois
    for recognition in recognitions:
TypeError: 'NoneType' object is not iterable

[ERROR] [1515516762.924176]: Robot exited with <class 'smach.exceptions.InvalidUserCodeError'>,Could not execute state 'DETECT_CROWD' of type '<challenge_spr_states.detect.DetectCrowd object at 0x7f1ab2fa0610>': Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/smach/state_machine.py", line 247, in _update_once
    self._remappings[self._current_label]))
  File "/home/amigo/ros/kinetic/repos/https_/github.com/tue-robotics/tue_robocup.git/challenge_spr/src/challenge_spr_states/detect.py", line 30, in execute
    detections = self.recognize(tries)
  File "/home/amigo/ros/kinetic/repos/https_/github.com/tue-robotics/tue_robocup.git/challenge_spr/src/challenge_spr_states/detect.py", line 59, in recognize
    faces = img_cutout(best_image, img_recognitions_to_rois(best_detection))
  File "/home/amigo/ros/kinetic/system/src/robot_skills/src/robot_skills/util/image_operations.py", line 15, in img_recognitions_to_rois
    for recognition in recognitions:
TypeError: 'NoneType' object is not iterable
,<traceback object at 0x7f1ab2f3a248>
[INFO] [1515516762.924547]: Torso cancelling all goals on close
[INFO] [1515516762.924901]: left arm cancelling all goals on all arm-related ACs on close
[INFO] [1515516762.925392]: right arm cancelling all goals on all arm-related ACs on close
Traceback (most recent call last):
  File "/home/amigo/ros/kinetic/system/src/challenge_spr/src/challenge_spr.py", line 84, in <module>
    startup(ChallengeSpeechPersonRecognition, challenge_name="challenge_spr")
  File "/home/amigo/ros/kinetic/system/src/robot_smach_states/src/robot_smach_states/util/startup.py", line 81, in startup
    outcome = executioner.execute()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/smach/state_machine.py", line 359, in execute
    container_outcome = self._update_once()
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/smach/state_machine.py", line 254, in _update_once
    + traceback.format_exc())
smach.exceptions.InvalidUserCodeError: Could not execute state 'DETECT_CROWD' of type '<challenge_spr_states.detect.DetectCrowd object at 0x7f1ab2fa0610>': Traceback (most recent call last):
  File "/opt/ros/kinetic/lib/python2.7/dist-packages/smach/state_machine.py", line 247, in _update_once
    self._remappings[self._current_label]))
  File "/home/amigo/ros/kinetic/repos/https_/github.com/tue-robotics/tue_robocup.git/challenge_spr/src/challenge_spr_states/detect.py", line 30, in execute
    detections = self.recognize(tries)
  File "/home/amigo/ros/kinetic/repos/https_/github.com/tue-robotics/tue_robocup.git/challenge_spr/src/challenge_spr_states/detect.py", line 59, in recognize
    faces = img_cutout(best_image, img_recognitions_to_rois(best_detection))
  File "/home/amigo/ros/kinetic/system/src/robot_skills/src/robot_skills/util/image_operations.py", line 15, in img_recognitions_to_rois
    for recognition in recognitions:
TypeError: 'NoneType' object is not iterable

--- a/challenge_spr/src/challenge_spr_states/detect.py
+++ b/challenge_spr/src/challenge_spr_states/detect.py
@@ -21,16 +21,14 @@ timeout = 10
 
 
 class DetectCrowd(smach.State):
+    """
+    Detect a crowd and describe it based on individual detections
+    """
     def __init__(self, robot):
         smach.State.__init__(self, outcomes=['succeeded', 'failed'], output_keys=['crowd_data'])
         self.robot = robot
 
     def execute(self, userdata=None):
-        """
-
-        :param userdata:
-        :return:
-        """
         tries = 3
         detections = self.recognize(tries)
 
@@ -41,9 +39,11 @@ class DetectCrowd(smach.State):
 
     def recognize(self, tries=3):
         """
-        Recognize people. Takes multiple images and 
+        Recognize people. Takes multiple images and takes the try with the most faces
         :param tries: number of tries
-        :return:
+        :type tries: int
+        :return: list of face properties
+        :rtype: list[image_recognition_msgs/FaceProperties]
         """
         number_of_people = 0
         best_image = None
@@ -70,12 +70,19 @@ class DetectCrowd(smach.State):
             faces = img_cutout(best_image, img_recognitions_to_rois(best_detection))
 
             rospy.loginfo('Calling Skybiometry...')
-
             return self.robot.perception.get_face_properties(faces=faces)
         else:
+            rospy.logerr("No faces detected during all {} tries".format(tries))
             return []
 
     def describe_crowd(self, detections):
+        """
+        Conversion from individual face properties to crowd properties
+        :param detections: list of face properties
+        :type detections: list[image_recognition_msgs/FaceProperties]
+        :return: crowd properties
+        :rtype: dict
+        """
         num_males = 0
         num_females = 0
         num_women = 0

--- a/challenge_spr/src/challenge_spr_states/detect.py
+++ b/challenge_spr/src/challenge_spr_states/detect.py
@@ -25,7 +25,12 @@ class DetectCrowd(smach.State):
         smach.State.__init__(self, outcomes=['succeeded', 'failed'], output_keys=['crowd_data'])
         self.robot = robot
 
-    def execute(self, userdata):
+    def execute(self, userdata=None):
+        """
+
+        :param userdata:
+        :return:
+        """
         tries = 3
         detections = self.recognize(tries)
 
@@ -34,7 +39,12 @@ class DetectCrowd(smach.State):
 
         return 'succeeded'
 
-    def recognize(self, tries):
+    def recognize(self, tries=3):
+        """
+        Recognize people. Takes multiple images and 
+        :param tries: number of tries
+        :return:
+        """
         number_of_people = 0
         best_image = None
         best_detection = None
@@ -56,11 +66,14 @@ class DetectCrowd(smach.State):
                 best_image = image
                 best_detection = face_rois
 
-        faces = img_cutout(best_image, img_recognitions_to_rois(best_detection))
+        if best_image:  # at least one face detected
+            faces = img_cutout(best_image, img_recognitions_to_rois(best_detection))
 
-        rospy.loginfo('Calling Skybiometry...')
+            rospy.loginfo('Calling Skybiometry...')
 
-        return self.robot.perception.get_face_properties(faces=faces)
+            return self.robot.perception.get_face_properties(faces=faces)
+        else:
+            return []
 
     def describe_crowd(self, detections):
         num_males = 0

--- a/robot_skills/src/robot_skills/perception.py
+++ b/robot_skills/src/robot_skills/perception.py
@@ -252,6 +252,7 @@ class Perception(RobotPart):
             rois = img_recognitions_to_rois(face_recognitions)
             faces = img_cutout(image, rois)
 
+        face_properties = []
         try:
             face_properties_response = self._face_properties_srv(faces)
             face_properties = face_properties_response.properties_array

